src/python/pants/engine/selectors.py

```diff
@@ -4,26 +4,31 @@
 import ast
 from dataclasses import dataclass
 from textwrap import dedent
-from typing import Any, Generator, Iterable, Tuple, Type, cast
+from typing import Any, Generator, Generic, Iterable, Optional, Tuple, Type, TypeVar, cast
 
 from pants.util.meta import frozen_after_init
 from pants.util.objects import TypeConstraint
 
 
+_Product = TypeVar("_Product")
```

Comment:
A docstring/comment for this would be good.

Reply:
A comment describing what in particular? This is fairly typical MyPy usage and we use TypeVars in several places currently.

Reply:
This formulation of generic parameters ("scoped to an entire file", rather than how they work in most languages "scoped to a particular class or function") makes it more important to explain what any particular generic parameter is/means, to avoid using the same name in a bunch of places that would dilute its meaning. This one seems to be used roughly as "the output type of a request for a Product"... so it might be `_OutputProduct`...?

Reply:
Totally agreed!